File: lexer/lexer.cpp
#include "lexer.h"
#include <sstream>
#include <iostream>

using std::cin;
using std::cout;
using std::stringstream;

// Constructor
Lexer::Lexer()
{
    // Inserts reserved words into the id's table
    id_table["true"] = Id{Tag::TRUE, "true"};
    id_table["false"] = Id{Tag::FALSE, "false"};
}

// Returns tokens to the parser
Token Lexer::Scan()
{
    // Skips spaces, tabs and new lines
    while (isspace(peek))
    {
        if (peek == '\n')
	{
	    line += 1;
	};
	peek = cin.get();
    };

    // Returns the numbers
    if (isdigit(peek))
    {
        int v = 0;


	do
	{
	    // Converts char to int
	    int n = peek - '0';
	    v = 10 * n + v;
	    peek = cin.get();
        }
	while (isdigit(peek));

	// DEBUG: shows the recognized token
	cout << "<num, " << v << "> ";

	// Returns the NUM token
	return Num{v};
    };

    // Returns the KW and id's
    if (isalpha(peek))
    {
        stringstream ss;

	do
	{
	    ss << peek;
	    peek = cin.get();
	}
	while (isalpha(peek));

	string s = ss.str();
	auto pos = id_table.find(s);

	// If the lexeme is already in the table
	if (pos != id_table.end())
	{
	    // DEBUG: shows the recognized token
	    switch (pos->second.tag)
	    {
	        case TRUE: cout << "<TRUE, " << pos->second.name << ">"; break;
		case FALSE: cout << "<FALSE, " << pos->second.name << ">"; break;
		default: cout << "<ID, " << pos->second.name << ">"; break;
	    };

	    // Returns the ID token
	    return pos->second;
	};

        // If the lexeme is not in the table
        Id new_id {Tag::ID, s};
	id_table[s] = new_id;

	// DEBUG
	cout << "<ID, " << new_id.name << ">";

	// Returns the id token
	return new_id;
    };

    // Operators and isolated alphanumeric characters
    Token t {peek};
    peek = ' ';
    
    // DEBUG
    cout << "<" << char(t.tag) << "> ";

    // Returns the token
    return t;
};

void Lexer::Start()
{
    // Simulates the parser
    while (peek != '\n')
    {
        Scan();
    };
};

File: python_lexer/lexer.py
class Tag:
    def __init__(self):
        self.ID = "ID"
        self.NUM = "NUM"
        self.TRUE = "TRUE"
        self.FALSE = "FALSE"

class Token:
    def __init__(self, tag):
        self.tag = tag

class Id(Token):
    def __init__(self, tag, name):
        self.tag = tag
        self.name = name
        super().__init__(self.tag)

class Num(Token):
    def __init__(self, value):
        self.value = value
        self.tag = Tag().NUM
        super().__init__(self.tag)

class Lexer:
    def __init__(self, source):

        self.source_code = source
        self.index = 0

        self.line = 1
        self.peek = self.source_code[self.index]

        self.id_table = {}
        self.id_table["false"] = Id(Tag().FALSE, "false")
        self.id_table["true"] = Id(Tag().TRUE, "true")

    def next_char(self):
        if len(self.source_code) > (self.index + 1):
            self.index += 1
            self.peek = self.source_code[self.index]
        else:
            self.peek = " "

    def scan(self):
        while self.peek.isspace():
            if self.peek == "\n":
                self.line += 1
            self.next_char()

        if self.peek.isdigit():
            v = 0
            n = int(self.peek)
            v = 10 * v + n
            self.next_char()

            while self.peek.isdigit():
                n = int(self.peek)
                v = 10 * v + n
                self.next_char()

            # DEBUG
            print(f"<NUM, {v}>")

            return Num(v)
        
        if self.peek.isalpha():
            s = self.peek
            self.next_char()

            while self.peek.isalpha():
                s += self.peek
                self.next_char()

            value = self.id_table.get(s)

            if value is not None:

                # DEBUG
                if s == "false":
                    print(f"<FALSE, {value.name}>")
                elif s == "true":
                    print(f"<TRUE, {value.name}>")
                else:
                    print(f"<ID, {value.name}>")

                return value

            new_id = Id(Tag().ID, s)
            self.id_table[s] = new_id

            # DEBUG
            print(f"<ID, {new_id.name}>")

            return new_id
        

        t = Token(self.peek)
        peek = ' '

        # DEBUG
        print(f"<{t.tag}>")

        return t

    def start(self):
        while self.peek != "\n":
            self.scan()

